File: Simple Shopping List/app/src/main/java/com/codeprecious/shoppinglistmvvm_hilt_rxjava/repository/Repository.java
package com.codeprecious.shoppinglistmvvm_hilt_rxjava.repository;


import androidx.lifecycle.LiveData;
import androidx.lifecycle.LiveDataReactiveStreams;

import com.codeprecious.shoppinglistmvvm_hilt_rxjava.room.ItemDao;
import com.codeprecious.shoppinglistmvvm_hilt_rxjava.room.ItemEntity;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.schedulers.Schedulers;


public class Repository {

    private ItemDao itemDao;

    @Inject()
    public Repository(ItemDao itemDao) {
        this.itemDao = itemDao;
    }

    public Completable insert(ItemEntity item) {
        return itemDao.insert(item);
    }

    public Completable delete(ItemEntity item) {
        return itemDao.delete(item);
    }

    public LiveData<List<ItemEntity>> getItem() {
        return LiveDataReactiveStreams.fromPublisher(itemDao.getAll()
                .subscribeOn(Schedulers.io()));
    }

}

File: FakeDownloadApp/app/src/main/java/com/codeprecious/fakedownloadapp/repository/Repository.kt
package com.codeprecious.fakedownloadapp.repository

import android.content.Context
import android.widget.Toast
import androidx.lifecycle.LiveData
import androidx.lifecycle.LiveDataReactiveStreams
import androidx.lifecycle.MutableLiveData
import com.codeprecious.fakedownloadapp.model.Data
import com.codeprecious.fakedownloadapp.retrofit.RetrofitApi
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import dagger.hilt.android.qualifiers.ApplicationContext
import io.reactivex.rxjava3.core.Flowable
import io.reactivex.rxjava3.schedulers.Schedulers
import java.io.IOException
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class Repository @Inject constructor(
    private val retrofitApi: RetrofitApi,
    @ApplicationContext private val context: Context
) {

    fun getData(): LiveData<List<Data>> {
        return LiveDataReactiveStreams.fromPublisher(
            retrofitApi.getFakeData().subscribeOn(Schedulers.io()).onErrorReturn {
                getDataFromJSON()
            }
        )
    }

    private fun getDataFromJSON(): List<Data> {
        val gson = Gson()
        val listData = object : TypeToken<List<Data>>() {}.type
        return gson.fromJson(getJSONDataFromAsset(), listData)
    }

    private fun getJSONDataFromAsset(): String? {
        return try {
            context.assets.open("getListOfFilesResponse.json")
                .bufferedReader().use {
                    it.readText()
                }
        } catch (e: IOException) {
            e.printStackTrace()
            null
        }
    }
}